### PR TITLE
Make AQL query optimizer create less query execution plan permutations

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Enterprise companion PR: https://github.com/arangodb/enterprise/pull/1471

Make AQL query optimizer create less query execution plan permutations

* Make AQL query optimizer create less query execution plan permutations when a query contains multiple occurrences of COLLECT statements under certain conditions:
  - COLLECT statements that are contained in a single loop that is driven by a traversal or an enumeration over an array now prefer the hashed collect variant.
  - if the hashed collect variant is explicitly requested via the `OPTIONS` clause in the COLLECT statement, it is always applied, even if the maximum number of query execution plans is already reached.

The overall goal of this change is to reduce the number of generated query execution plans for queries that contains lots of COLLECT statements, in order to reduce the total time spent for query optimization.

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [x] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.12.0: -
  - [ ] Backport for 3.11: -
  - [ ] Backport for 3.10: -

#### Related Information

- [ ] Docs PR: 
- [x] Enterprise PR: https://github.com/arangodb/enterprise/pull/1471
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/Aql/OptimizerRules.cpp [hunk 1]
```diff
@@ -1943,101 +1943,107 @@ void arangodb::aql::specializeCollectRule(Optimizer* opt,
...
+      if (ExecutionNode const* loop = collectNode->getLoop();
+          loop != nullptr && loop->getLoop() == nullptr &&
+          (loop->getType() == EN::ENUMERATE_LIST ||
+           loop->getType() == EN::TRAVERSAL ||
+           loop->getType() == EN::ENUMERATE_PATHS ||
+           loop->getType() == EN::SHORTEST_PATH)) {
+        // if the COLLECT is contained inside a single loop, and the loop is an
+        // enumeration over an array (in contrast to an enumeration over a
+        // collection/view, then prefer the hashed collect variant. this is
+        // because the loop output is unlikely to be sorted in any way.
```
Comment (MBkkt) [inline arangod/Aql/OptimizerRules.cpp:1961]: I thought we had some optimization for sort `ENUMERATE_LIST` in case of optimization time known list

Comment (jsteemann) [reply arangod/Aql/OptimizerRules.cpp:1961]: The only optimization I know of that optimizes away sorts is the one that removes sorts for inputs which are produced by a sorted index. Maybe there exists also such optimization for inputs produced by a sorted view. 
But I am not aware of an optimization that checks the sortedness of an `ENUMERATE_LIST` input that is known at optimization time. We could write one though :sweat_smile:

**Review by maierlars — APPROVED** [review]

File: CHANGELOG [hunk 2]
```diff
@@ -1,6 +1,16 @@
...
 
+* Make AQL query optimizer create less query execution plan permutations when
+  a query contains multiple occurrences of COLLECT statements under certain
+  conditions:
+  - COLLECT statements that are contained in a single loop that is driven by a
+    traversal or an enumeration over an array now prefer the hashed collect
+    variant.
+  - if the hashed collect variant is explicitly requested via the `OPTIONS`
+    clause in the COLLECT statement, it is always applied, even if the maximum
+    number of query execution plans is already reached.
```
Comment (hkernbach) [inline CHANGELOG:12]: We could add an example here how those OPTIONS look like in an AQL query.

File: arangod/Aql/ExecutionNode/CollectNode.cpp [hunk 3]
```diff
@@ -141,10 +137,19 @@ void CollectNode::doToVelocyPack(VPackBuilder& nodes,
...
   nodes.add(VPackValue("collectOptions"));
   _options.toVelocyPack(nodes);
+
+  // this attribute is not read anymore since 3.12.1 and is only
+  // kept for compatibility with older versions. it can be removed
+  // entirely in a future version
+  nodes.add("specialized", VPackValue(_options.isFixed()));
+
+  // TODO: remove this attribute too - it is not read back since
+  // 3.12.1
```
Comment (hkernbach) [inline arangod/Aql/ExecutionNode/CollectNode.cpp:149]: There should be a JIRA card for this. If this exists, ignore my comment.

File: arangod/Aql/OptimizerRules.cpp [hunk 4]
```diff
@@ -1933,7 +1933,7 @@ void arangodb::aql::specializeCollectRule(Optimizer* opt,
   for (auto const& n : nodes) {
     auto collectNode = ExecutionNode::castTo<CollectNode*>(n);
 
-    if (collectNode->isSpecialized()) {
+    if (collectNode->isFixedMethod()) {
       // already specialized this node
```
Comment (hkernbach) [inline arangod/Aql/OptimizerRules.cpp:1937]: This comment might now need an adjustment?

File: arangod/Aql/OptimizerRules.cpp [hunk 5]
```diff
@@ -1943,101 +1943,107 @@ void arangodb::aql::specializeCollectRule(Optimizer* opt,
...
+        parent->replaceDependency(collectNode, sortNode);
 
         modified = true;
         continue;
       }
 
-      // create a new plan with the adjusted COLLECT node
-      std::unique_ptr<ExecutionPlan> newPlan(plan->clone());
+      // are we allowed to generate additional plans?
+      if (!opt->runOnlyRequiredRules(1)) {
```
Comment (hkernbach) [inline arangod/Aql/OptimizerRules.cpp:1995]: This sounds (comment) like the method could get a better fitting name.

**Review by hkernbach — COMMENTED** [review]
Left some comments. Might be all minor except the one related to the duplication of "method" sets in the node. Apart from that LGTM.

File: CHANGELOG (hunk 2 above)
Comment (jsteemann) [reply CHANGELOG]: Addressed

File: arangod/Aql/ExecutionPlan.cpp [hunk 6]
```diff
@@ -994,13 +994,16 @@ CollectNode* ExecutionPlan::createAnonymousCollect(
...
-      nullptr, nullptr, std::vector<std::pair<Variable const*, std::string>>{},
-      _ast->variables()->variables(false), true);
+      this, nextId(), options, std::move(groupVariables),
+      std::move(aggregateVariables), nullptr, nullptr,
+      std::vector<std::pair<Variable const*, std::string>>{},
+      _ast->variables()->variables(false));
 
-  en->aggregationMethod(CollectOptions::CollectMethod::DISTINCT);
-  en->specialized();
+  en->aggregationMethod(CollectOptions::CollectMethod::kDistinct);
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/Aql/ExecutionPlan.cpp]: Fixed

File: arangod/Aql/OptimizerRules.cpp (hunk 4 above)
Comment (jsteemann) [reply arangod/Aql/OptimizerRules.cpp]: Addressed

File: arangod/Aql/OptimizerRules.cpp (hunk 5 above)
Comment (jsteemann) [reply arangod/Aql/OptimizerRules.cpp]: Did not rename the method, but removed the potentially confusing `1` parameter.

**Review by hkernbach — APPROVED** [review]

**Pull request merged**